File: 820-find-eventual-safe-states/find-eventual-safe-states.cpp
class Solution {
public:
    vector<int> eventualSafeNodes(vector<vector<int>>& graph) {
        int v=graph.size();
        vector<vector<int>>adjrev(v);
        for(int i=0;i<v;i++){
            for(auto it:graph[i]){
                adjrev[it].push_back(i);
            }
        }
        vector<int>indegree(v,0);
        for(int i=0;i<v;i++){
            for(auto it:adjrev[i]){
                indegree[it]++;
            }
        }
        queue<int>q;
        vector<int>ans;
        for(int i=0;i<v;i++){
            if(indegree[i]==0){
                q.push(i);
            }
        }
        while(!q.empty()){
            int node=q.front();
            q.pop();
            ans.push_back(node);

            for(auto it:adjrev[node]){
                indegree[it]--;
                if(indegree[it]==0){
                    q.push(it);
                }
            }
        }
        sort(ans.begin(),ans.end());
        return ans;
    }
};

File: 820-find-eventual-safe-states/find-eventual-safe-states.py
from typing import List

class Solution:
    def eventualSafeNodes(self, graph: List[List[int]]) -> List[int]:
        n = len(graph)
        vis = [0] * n   

        def dfs(node: int) -> bool:
            if vis[node] != 0:
                return vis[node] == 2  

            vis[node] = 1
            for nei in graph[node]:
                if vis[nei] == 1 or not dfs(nei):
                    return False
            vis[node] = 2 
            return True

        ans = []
        for i in range(n):
            if dfs(i):
                ans.append(i)
        return ans
